File: greedy/knapsack.cpp
#include<iostream>
#include<map>

using namespace std;


int knapsack(map<int,int> wp, int k)
{
    int pro=0;
    for (auto  it = wp.rbegin(); it != wp.rend(); it++)
    {
        wp.insert({it->first/it->second, it->second});
        wp.erase(it->first);
        
        
    }
    for (auto  it = wp.rbegin(); it != wp.rend(); it++)
    {
        cout<<it->first<<" "<<it->second<<endl;
        
    }
    



return 1;

}
int main(int argc, char const *argv[])
{
    int n,w,p,k;
    cout<<"Enter size: ";
    cin>>n;
    map<int,int> wp;
    for (int i = 0; i < n; i++)
    {
        cout<<"Enter weight: ";
        cin>>w;
        cout<<"Enter profit: ";
        cin>>p;
        wp.insert({p,w});
    }
    cout<<"Enter size pf knapsack: ";
    cin>>k;

    knapsack(wp,k);    
    return 0;
}

File: greedy/knapsack.py
n = int(input("Enter number: "))
def knapsack(wp,k):
    for i in range(len(wp)):
        wp[i][1] = wp[i][1]/wp[i][0]

    wp.sort(key = lambda x: x[1],reverse=True)
    print(wp)
    pro,i = 0,0
    while k>0 and i<=len(wp)-1:
        if k > wp[i][0]:
            pro+= wp[i][0]*wp[i][1]
            k-= wp[i][0]
        else:
            pro += k*wp[i][1]

        i+=1

    print(pro)








wp = list()
for i in range(n):
    a = list()
    a.append(int(input("Enter weight: ")))
    a.append(int(input("Enter profit: ")))
    wp.append(a)
k = int(input("Enter size of knapsack: "))
mx = knapsack(wp,k)